pkg/codegen/hcl2/binder_schema.go

```diff
@@ -198,3 +198,52 @@ func (b *binder) schemaTypeToType(src schema.Type) (result model.Type) {
 		}
 	}
 }
+
+// GetSchemaForType extracts the schema.Type associated with a model.Type, if any.
+//
+// The result may be a *schema.UnionType if multiple schema types are associaged with the input type.
+func GetSchemaForType(t model.Type) (schema.Type, bool) {
+	switch t := t.(type) {
+	case *model.ObjectType:
+		if len(t.Annotations) == 0 {
+			return nil, false
+		}
+		for _, a := range t.Annotations {
+			if t, ok := a.(schema.Type); ok {
+				return t, true
+			}
+		}
+		return nil, false
+	case *model.OutputType:
+		return GetSchemaForType(t.ElementType)
+	case *model.PromiseType:
+		return GetSchemaForType(t.ElementType)
+	case *model.UnionType:
+		schemas := codegen.Set{}
+		for _, t := range t.ElementTypes {
+			if s, ok := GetSchemaForType(t); ok {
+				if s, ok := s.(*schema.UnionType); ok {
+					for _, s := range s.ElementTypes {
+						schemas.Add(s)
+					}
+				} else {
+					schemas.Add(s)
```

Comment:
`s` is overridden here to the value of `UnionType` conversion, which will be `nil`. I think we should change the above `if` not to redeclare `s` to add the original result of `GetSchemaForType` to `schemas`.